### ISSUE TITLE
Training on Notebook

### ISSUE THREAD

**Issue by muzahidai** [black0017/3D-GAN-pytorch#1]
Thank you so much for sharing such nice repositories. I have run the notebook on colab successfully. I am interested to train the model on colab. could you please guide me how to train 3DGAN on colab and visualize the generated 3D model by the generator? Thank you once again.

**Comment by black0017** [comment]
Hello, thank you for your kind comment :)

I am leaving here a quick recap of an example training code flow for future reference.
You need to plug in the data that you want to train the model and save the data that you want to keep track of the training progress. If it usefull let me know so I can add it in the project files. It is a little bit high level, based on the code clean up. 

I used this code on private data :) 

All the best
N.A.

```python
import torch
from torch import nn, optim

import train_gans as train


def example_train(discriminator, generator, data, args):
    # init weight normal distr.
    generator.apply(init_weights_normal)
    discriminator.apply(init_weights_normal)

    if args.cuda:
        discriminator.cuda()
        generator.cuda()
        # 10 test samples
        test_noise = noise(10).cuda()

    # LOSSES
    loss_d = nn.BCELoss()
    loss_g = nn.BCELoss()

    # OPTIMIZERS

    d_optimizer = optim.Adam(discriminator.parameters(), args.lr_d, betas=args.betas)
    g_optimizer = optim.Adam(generator.parameters(), args.lr_g, betas=args.betas)

    # Data loader
    data_loader = torch.utils.data.DataLoader(data, batch_size=args.batch_size, shuffle=True)
    num_batches = len(data_loader)

    for epoch in range(args.start_epoch, args.num_epochs):
        for n_batch, sample_subject in enumerate(data_loader):
            iteration = epoch * num_batches + n_batch

            real_data, _ = sample_subject

            # 1. Train Discriminator
            latent_space_data = noise(args.batch_size)
            if args.cuda:
                latent_space_data = latent_space_data.cuda()
                real_data = real_data.cuda()

                fake_data = generator(latent_space_data).detach()

                d_error, d_pred_real, d_pred_fake = train.train_discriminator(
                    discriminator, d_optimizer, real_data, fake_data,
                    loss_d, epoch, args.cuda)

            # 2. Train Generator
            latent_space_data = noise(args.batch_size)
            if args.cuda:
                latent_space_data = latent_space_data.cuda()

            fake_data = generator(latent_space_data)
            g_error = train.train_generator(discriminator, g_optimizer, fake_data, loss_g, args.cuda)  # Log batch error

            # save the stats that you want here :)


def init_weights_normal(m):
    classname = m.__class__.__name__
    if classname.find('Conv') != -1 or classname.find('BatchNorm') != -1:
        m.weight.data.normal_(0.00, 0.02)


def noise(size, latent_dim=200):
    '''
    Generates a 1-d vector of gaussian sampled random values
    '''
    n = torch.randn(size, latent_dim)
    return n


# Functions that you ll definately need :)
import os, shutil


def save_checkpoint_gan(state_d, state_g, is_best, path, prefix, filename='checkpoint.pth.tar'):
    prefix_save = os.path.join(path, prefix)
    name_d = prefix_save + '_D_' + filename
    name_g = prefix_save + '_G_' + filename
    torch.save(state_d, name_d)
    torch.save(state_g, name_g)
    if is_best:
        shutil.copyfile(name_d, prefix_save + '_model_D_best.pth.tar')


def load_checkpoints_gan(model_d, model_g, path_d, path_g):
    checkpoint_d = torch.load(path_d)
    checkpoint_g = torch.load(path_g)
    model_d.load_state_dict(checkpoint_d['state_dict'])
    model_g.load_state_dict(checkpoint_g['state_dict'])
    return model_d, model_g

```

**Comment by sajmani4913** [comment]
Thank you for sharing this @black0017. May I know if you have used the eval function somewhere in the code? I don't see the "eval" in the whole repository. I want to train the model for my research purposes with different optimizers, loss functions, and parameter values. Still, apart from your code in the comments, I do not see anything mentioned by authors on how to train and run a code from scratch, not just using the trained weights. Any help is appreciated. :) @zck119

**Comment by ZahraMESBAHI** [comment]
Hello @muzahidai , hope you're doing well. I am a beginner am confused how to improve my knowledge, and how to get started. I saw your comment and I want to ask you about the data you used to run this code. Also I had issues with my codes and I don't know where can find people who I can ask them and help me. Please give a help and advice. 

Thank you in advance
